python_tests/test_csf.py

```diff
@@ -21,23 +21,23 @@ def run_csf_test(keys, values):
 
 
 @pytest.mark.unit
-def test_csf():
+def test_csf_int():
     keys = [f"key{i}" for i in range(1000)]
     values = [i for i in range(1000)]
-    run_csf_test(keys, values)
+    run_csf_int_test(keys, values)
 
 
 @pytest.mark.unit
-def test_csf_byte_strings():
+def test_csf_byte_string_keys():
     keys = [f"key{i}".encode("utf-8") for i in range(1000)]
     values = [i for i in range(1000)]
-    run_csf_test(keys, values)
+    run_csf_int_test(keys, values)
 
 
 @pytest.mark.unit
 def test_csf_char_10():
     keys = [f"key{i}".encode("utf-8") for i in range(1000)]
-    values = [list(f"value{i}".ljust(10)[:10]) for i in range(1000)]
+    values = [str(f"value{i}".ljust(10)[:10]) for i in range(1000)]
```

Comment:
This test is failing locally but not for the reason I first thought... Somehow pybind is able to convert these string values to  `std::array<char, 10> ` but the output of `csf.query(key)` is still a list of characters:

`FAILED ../python_tests/test_csf.py::test_csf_char_10 - AssertionError: assert ['v', 'a', 'l', 'u', 'e', '0', ...] == 'value0    '`

I'm not sure if you know of some pybind magic out there that's cleaner but the only thing I can think of is to wrap the csf `query` method to check the template and return an std::string whenever the type is a char array. What do you think?

Reply:
With my compiler on my machine, pybind correctly typecasts the return value to a string. But holy crap! The reason why this happens is just insane!

Pybind11 supports [a ton of different type conversion rules](https://pybind11.readthedocs.io/en/stable/advanced/cast/overview.html#list-of-all-builtin-conversions), but `char[]` isn't one of them. Seems like your compiler and my compiler are choosing "intermediate" C++ types that convert to different Python types. Totally nuts.

We can try expressing this as a `std::array<T, Size>`, which has stable type conversion rules in pybind. `char*` with a fixed-size underlying buffer is another option, but we'd have to null-terminate everything. Worst-case, we can wrap the CSFCharX classes like this (but I dislike this hack):
1. In the PythonBindings.cc, replace CSFCharX with _CSFCharX
2. In the __init__.py, wrap a _CSFCharX with a post-conversion on the query, like this:
```python
class CSFCharX:
   def __init__(self, csf):
      self._csf = # of type _CSFCharX
   def query(self, q):
      return str(self._csf.query(q))
# Passthrough all the other methods of _CSFCharX
```

Reply:
Fixed via wrapper class.